### PR TITLE
Update parameter graph to full DAG

### PR DESCRIPTION
**Pull request by ConnorStoneAstro**
Updated parameter system allows for far more complex parameter graphs with a great deal of interaction and also user defined functions in between parameters. This allows for a number of powerful fitting tasks to be performed. These include:
- fitting an astrometric solution for all objects in an image
- imposing a functional form on parameters, such as a spatially varying PSF with spline interpolation
- linking parameters with different names (serious limitation of old system)
- Adding complex equality constraints (e.g. x must be y+5) instead of only strict equality

All tests run except for fitting methods which require significant updates to use the new parameter system.

### CODE REVIEW

**Comment by ConnorStoneAstro** [comment]
Ok, The full system seems to be working! Next I'm going to update the tutorials and add some demos of how to use the new system.

**Comment by ConnorStoneAstro** [comment]
Everything seems to be online. A new tutorial has been made and added to the `AstroPhot-tutorials` repo: https://github.com/Autostronomy/AstroPhot-tutorials/tree/params

Looks ready to go!

**Comment by ConnorStoneAstro** [comment]
Hopefully this will be the last giant 1000+ line update for a while!

File: astrophot/models/psf_model.py [hunk 1]
```diff
@@ -60,15 +61,11 @@ def __init__(self, *args, **kwargs):
...
-            )
-        if parameters["flux"].uncertainty is None:
-            parameters["flux"].set_uncertainty(
-                torch.abs(parameters["flux"].value) * 1e-2, override_locked=True
-            )
+        with Param_Unlock(parameters["flux"]), Param_SoftLimits(parameters["flux"]):
+            if parameters["flux"].value is None:
+                parameters["flux"].value = torch.log10(torch.abs(torch.sum(target_area.data)) / target.pixel_area)
+            if parameters["flux"].uncertainty is None:
+                parameters["flux"].uncertainty = torch.abs(parameters["flux"].value) * 1e-2
```
Comment (wmwv) [inline astrophot/models/psf_model.py]: I appreciate that this line isn't really changed in this PR.  But I would encourage a short note saying something like you assume a 1% uncertainty floor if uncertainty is not specified, and then defining a named variable to hold `1e-2`.  Bare numbers without documentation make me worried; assigning them to a named variable can be a nice way of documenting what they are.

File: astrophot/models/warp_model.py [hunk 2]
```diff
@@ -84,21 +85,16 @@ def initialize(self, target=None, parameters=None, **kwargs):
...
-
-        if parameters["q(R)"].value is None:
-            parameters["q(R)"].set_value(
-                np.ones(len(parameters["q(R)"].prof)) * 0.9, override_locked=True
-            )
+        with Param_Unlock(parameters["PA(R)"]), Param_SoftLimits(parameters["PA(R)"]):
+            if parameters["PA(R)"].value is None:
+                parameters["PA(R)"].value = np.zeros(len(parameters["PA(R)"].prof)) + target.north
+            if parameters["q(R)"].value is None:
+                parameters["q(R)"].value = np.ones(len(parameters["q(R)"].prof)) * 0.9
```
Comment (wmwv) [inline astrophot/models/warp_model.py:103]: Similar question/suggestion about what is 0.9, and documenting it at least in a variable name.  And a similar recognition that this is not a real change in this PR, but the line was touched, so I take the liberty of commenting.

File: tests/test_fit.py [hunk 3]
```diff
@@ -292,41 +294,41 @@ def test_lm_creation(self):
...
-        )
-        self.assertTrue(
-            torch.all(torch.isclose(jac, torch.ones_like(jac))),
-            "Constraint Jacobian should produce correct derivatives",
-        )
-        self.assertTrue(
-            torch.all(torch.isfinite(samp)),
-            "Constraint sample should produce real numbers",
-        )
+#     def test_lm_constraint(self):
```
Comment (wmwv) [inline tests/test_fit.py]: Comment on why these were commented out and why a future person might want to reactivate them or retrieve the logic.

File: tests/test_model.py [hunk 4]
```diff
@@ -29,26 +29,26 @@ def test_AstroPhot_Model(self):
...
-                "Re": 5,
-                "Ie": 1,
-            },
-            target=target,
-        )
-        rep = model.parameters.get_vector(as_representation = True)
-        nat = model.parameters.get_vector(as_representation = False)
-        self.assertTrue(torch.all(torch.isclose(rep, model.parameters.transform(nat, to_representation = True))), "transform should map between parameter natural and representation")
-        self.assertTrue(torch.all(torch.isclose(nat, model.parameters.transform(rep, to_representation = False))), "transform should map between parameter representation and natural")
+    # def test_basic_model_methods(self):
```
Comment (wmwv) [inline tests/test_model.py]: Remove completely or add comment about why it should be saved and what is useful about it.

File: astrophot/param/param_context.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+__all__ = ("Param_Unlock", "Param_SoftLimits", "Param_Mask")
+
+class Param_Unlock:
+    """Temporarily unlock a parameter.
+
+    Context manager to unlock a parameter temporarily. Inside the
+    context, the parameter will behave as unlocked regardless of its
+    initial condition. Upon exiting the context, the parameter will
+    return to it's previous locked state regardless of any changes
```
Comment (wmwv) [inline astrophot/param/param_context.py]:
```suggestion
    return to its previous locked state regardless of any changes
```

File: astrophot/param/param_context.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+        if self.param is None:
+            Node.global_unlock = False
+        else:
+            self.param.locked = self.original_locked
+
+class Param_SoftLimits:
+    """
+    Temporarily allow writing parameter values outside limits.
+
+    Values outside the limits will be quietly shift until they are within the boundaries of the parameter limits.
```
Comment (wmwv) [inline astrophot/param/param_context.py]: "quietly shifted"?
Which boundary will they be shifted to in the case of cyclical constraints, such as PA?

File: astrophot/param/parameter.py [hunk 7]
```diff
@@ -0,0 +1,462 @@
...
+            if S == 1:
+                return np.array((self.name,))
+            return np.array(tuple(f"{self.name}:{i}" for i in range(self.size)))
+        flat = self.flat(include_locked = False, include_links = False)
+        return np.concatenate(tuple(node.names() for node in flat.values()))
+    
+    def vector_values(self):
+        """The vector representation is for values which correspond to
+        fundamental inputs to the parameter DAG. Since the DAG may
+        have linked nodes, or functions which produce values derrived
```
Comment (wmwv) [inline astrophot/param/parameter.py]:
```suggestion
        have linked nodes, or functions which produce values derived
```

**Review by wmwv — COMMENTED** [review]
Overall, looks fine.  I'm going to try to run stuff tomorrow before signing off on the PR.

Overall, I'd encourage you to make sure all the lines of the new code of `parameter.py and `param_context.py` are covered by tests.

**Comment by ConnorStoneAstro** [comment]
just pushed some updates to docstrings. I'll go through your comments in the morning. Thanks for reviewing so quickly!

File: astrophot/models/psf_model.py (hunk 1 above)
Comment (ConnorStoneAstro) [reply astrophot/models/psf_model.py]: That makes sense to me, I think I did this just as a stop gap in case initial uncertainties aren't provided. I can make it more explicit what the meaning of the `1e-2` is.

File: astrophot/models/warp_model.py (hunk 2 above)
Comment (ConnorStoneAstro) [reply astrophot/models/warp_model.py:103]: In this case I picked `0.9` as a heuristic. I want to start the axis ratio profile at some neutral position, `1` would be ideal since it would represent no change. However I picked 0.9 so that it doesn't hit the parameter boundary of 1 and so there is some signal in the optimizer as to what happens when the `q(R)` value is adjusted. If I started it at 0.999 it might get stuck and take a while to get the parameter out of that extreme border value. 

I'll put something to this effect in the docs

File: tests/test_fit.py (hunk 3 above)
Comment (ConnorStoneAstro) [reply tests/test_fit.py]: Oh, good catch. I can get rid of them now since the new constraint system is much better and this test is obsolete.

File: tests/test_model.py (hunk 4 above)
Comment (ConnorStoneAstro) [reply tests/test_model.py]: Oh, I should add this test back in. It's a good simple test of the transofrmations

File: astrophot/param/param_context.py (hunk 6 above)
Comment (ConnorStoneAstro) [reply astrophot/param/param_context.py]: "quietly shifted" just means no errors or warnings are thrown. For cyclical constraints there is no need to shift, I just use modulo `%` to wrap it back into the range. I'll add to the docs

**Comment by ConnorStoneAstro** [comment]
Ok, updated for the comments, going to add more tests next.

**Comment by ConnorStoneAstro** [comment]
Ok, new tests reach almost every line in the new param DAG stuff! Looks like this is ready to merge.
